### ISSUE TITLE
Shards wont get assigned after cluster restart

### ISSUE THREAD

**Issue by barton-im** [elastic/elasticsearch#21329]
Elasticsearch 1.7.1

OS: Debian 7

I have several indices in my cluster that got only partially assigned (some of the shards are normally assigned where the other shards are stuck in unassigned state). The cluster restart doesnt help, after restart always the same shards are not assigned. I got to this state when I was running out of space (had to raise the watermark for shard allocation) and installed new disks on the machines in a rolling fashion. But some of the shards got stuck still not being assigned, so I tried to disable the free space allocation altogether and restarted the cluster after which some of the shards came back but still some didnt. 
Now I am in a state where I see the data in the shards that are not assigned. I tried to check some with CheckIndex tool that says that everything is OK with that shard. I tried to look for .recovering files in the translog but there are none. I tried to change the replication and close open indices but in the logs I dont even see a trace of the ES trying to allocate the failed shard. 
I tried the reroute assign API, but without the allow_primary=true it just says the 400 error and if I allow the primary it deletes the existing shard with empty one so that is a no go for me. The reroute?explain says that the shards are "CLUSTER_RECOVERED".
I tried to mess around with the data dir (removing the " _state" and "translog" directories for the index and shard and restart the node but with no success.

Is there a way to somehow "reset" the ES metadata for a given shard so I dont loose the actual lucene data in the shard? Seems to me like some inconsistent ES state for the given shard(s).

**Comment by barton-im** [comment]
So, I have managed to bring my data back. However a lot of hoops to jump through. I have investigated the log on the master and found some traces of the problematic shards, notably this line was repeating:

[....] not allocating, number_of_allocated_shards_found [0], required_number [1]

note that this line appeared only on the master node, not on the node that actually carried the data. After few google searches I stumbled upon a setting for a cluster/index - index.recovery.initial_shards that allows to do recovery only in case a predefined number of replicas are already present. In my case, I needed that number to be 0 since none of the shard replicas was active (sometimes I had the index replicated 2x sometimes the replication was 0), in any case, I used the change setting API and set this to 0 on one testing index. It assigned the shard immediately, so I started changing this setting for other indices, one by one. However, on some of the indices the assign attempt was failing and the shards were going in loop from unassigned, to assign attempt on random nodes. I have tried to close and open the index bu that didnt help. The only thing that finally help to assign the most problematic shards was to locate the node where the data was physically stored and restart that node. 

So my problem is solved, however, some lessons could be learned from this exercize: the ES sometimes gets into incosistent state during allocations and there is (at least not to my knowledge) much tools to debug that. So I hope this will be fixed in some future versions (I know that there has been a long way already gone since I didnt have to do anything with the physical files which was needed to troubleshoot some problems with teh pre-1 versions).

The error that reported by ES while assigning the most problematic shards was:

sending shard started for [XXXXX][2], node[zMucPFz0RJuhsSVqgE6jJg], [P], s[INITIALIZING], unassigned_info[[reason=ALLOCATION_FAI
LED], at[2016-11-04T12:48:58.087Z], details[shard failure [failed recovery][IndexShardGatewayRecoveryException[[XXXXX][2] failed to fetch index version after copying it over]; nested: IndexShardGatewayRecov
eryException[[large-product-db-nm-20160518][2] shard allocated for local recovery (post api), should exist, but doesn't, current files: []]; nested: IndexNotFoundException[no segments* file found in store(least_used[rate_limited(
default(mmapfs(XXXXX),niofs(XXXXX)), type=MERGE, rate=20.0), rate_limited(defau
lt(mmapfs(XXXXX),niofs(XXXXX)), type=MERGE, rate=20.0)]): files: []]; ]]], inde
xUUID [ogg1XgLjTzyWlgeX56uycw], reason [after recovery from gateway]

**Comment by barton-im** [comment]
So, I have managed to bring my data back. However a lot of hoops to jump through. I have investigated the log on the master and found some traces of the problematic shards, notably this line was repeating:

[....] not allocating, number_of_allocated_shards_found [0], required_number [1]

note that this line appeared only on the master node, not on the node that actually carried the data. After few google searches I stumbled upon a setting for a cluster/index - index.recovery.initial_shards that allows to do recovery only in case a predefined number of replicas are already present. In my case, I needed that number to be 0 since none of the shard replicas was active (sometimes I had the index replicated 2x sometimes the replication was 0), in any case, I used the change setting API and set this to 0 on one testing index. It assigned the shard immediately, so I started changing this setting for other indices, one by one. However, on some of the indices the assign attempt was failing and the shards were going in loop from unassigned, to assign attempt on random nodes. I have tried to close and open the index bu that didnt help. The only thing that finally help to assign the most problematic shards was to locate the node where the data was physically stored and restart that node. 

So my problem is solved, however, some lessons could be learned from this exercize: the ES sometimes gets into incosistent state during allocations and there is (at least not to my knowledge) much tools to debug that. So I hope this will be fixed in some future versions (I know that there has been a long way already gone since I didnt have to do anything with the physical files which was needed to troubleshoot some problems with teh pre-1 versions).

The error that reported by ES while assigning the most problematic shards was:

sending shard started for [XXXXX][2], node[zMucPFz0RJuhsSVqgE6jJg], [P], s[INITIALIZING], unassigned_info[[reason=ALLOCATION_FAI
LED], at[2016-11-04T12:48:58.087Z], details[shard failure [failed recovery][IndexShardGatewayRecoveryException[[XXXXX][2] failed to fetch index version after copying it over]; nested: IndexShardGatewayRecov
eryException[[large-product-db-nm-20160518][2] shard allocated for local recovery (post api), should exist, but doesn't, current files: []]; nested: IndexNotFoundException[no segments* file found in store(least_used[rate_limited(
default(mmapfs(XXXXX),niofs(XXXXX)), type=MERGE, rate=20.0), rate_limited(defau
lt(mmapfs(XXXXX),niofs(XXXXX)), type=MERGE, rate=20.0)]): files: []]; ]]], inde
xUUID [ogg1XgLjTzyWlgeX56uycw], reason [after recovery from gateway]
